Step5 - 로또(수동)

Kelicia91 opened the pull request.

Kelicia91 committed:
docs: 로또 수동 구매를 위한 기능 목록 추가

Kelicia91 committed:
feat: LottoStore.sell() - 수동 판매 추가

Kelicia91 committed:
refactor: gardening

Kelicia91 committed:
refactor: LottoNumber.from - 매번 객체 생성하지 않고 캐싱된 객체 반환

Kelicia91 committed:
refactor: rename methods, vars

Kelicia91 committed:
refactor: gardening

Kelicia91 committed:
feat: View - 수동 복권 구매 입출력 추가

Kelicia91 committed:
fix: LottoApp - 수동 복권 구매 추가

- refactor: 복권 구매 제어 코드가 많아짐에 따라 LottoBuyer로 분리
- refactor: 구매한 복권 출력시 여백 처리를 view로 이전

Kelicia91 committed:
refactor: LottoPayment class 분리

Kelicia91 committed:
refactor: 로또(수동) 갯수를 ManualLottoAmount 클래스로 분리

Kelicia91 committed:
fix: LottoStore.sell 인자에 LottoPayment 적용

Kelicia91 committed:
fix: View - ManualLottoAmount 적용

Kelicia91 committed:
fix: LottoBuyer - LottoPayment, ManualLottoAmount 적용

minsungryu commented on src/main/java/LottoApp.java:
```diff
@@ -2,17 +2,16 @@
 	public void run() {
-		final List<Lotto> lottos = tryBuyLottos();
-		view.outBoughtLotto(lottos);
+		final List<Lotto> lottos = buyer.buy();
```
제가 전체적인 구조를 그려보진 않았지만 `LottoBuyer`나 `LottoStore`가 어느 하나에 의존한다기보단 
`buyer.buyAt(store)`이나 `store.sellTo(buyer)`처럼 메서드 호출시 인자로 넘기는 방법도 있을 것 같은데 어떻게 생각하시나요?

Kelicia91 replied:
buyer 를 buyLottosController 로 재명명 했구요, `buyLottosController.buyLottosAt(lottoStore)` 요렇게 호출하는 구조로 변경했어요!

minsungryu commented on src/main/java/LottoBuyer.java:
```diff
@@ -0,0 +1,84 @@
+
+	public LottoBuyer(LottoStore lottoStore) {
+		this.store = lottoStore;
+		this.view = new View();
```
`View`가 생성자 내부에서 초기화되는 것보다는 주입을 받는 쪽이 좋지 않을까요?
`LottoApp`에서도 새로 인스턴스를 생성하는 것 같더라구요

Kelicia91 replied:
`LottoApp`이 main controller 고, `BuyLottosController`가 sub controller 격이라 app 에서 view 하나만 생성하고 sub controller 생성시 view를 주입해주는 방식으로 변경했어요!

minsungryu commented on src/main/java/LottoNumber.java:
```diff
@@ -1,10 +1,18 @@
 	public final static int MIN_INCLUSIVE_NUMBER = 1;
 	public final static int MAX_INCLUSIVE_NUMBER = 45;
 
+	private final static Map<Integer, LottoNumber> CACHED_LOTTO_NUMBERS =
```
캐싱도 적용해주셨네요 👍

minsungryu commented on src/main/java/LottoPayment.java:
```diff
@@ -0,0 +1,37 @@
+
+	private final int numOfLottosCanBuy;
+
+	private LottoPayment(int KRW) {
```
https://rules.sonarsource.com/java/tag/convention/RSPEC-117

컨벤션에 따라 파라미터 이름을 소문자로 변경하면 좋을 것 같습니다.

Kelicia91 replied:
수정햇어요!

minsungryu commented on src/main/java/LottoPayment.java:
```diff
@@ -0,0 +1,37 @@
+
+	private LottoPayment(int KRW) {
+		validate(KRW);
+		this.numOfLottosCanBuy = KRW / Lotto.PRICE_KRW;
```
생성자에 로직이 있는 것 보다는 메서드 호출 시점에 계산하는 것은 어떨까요?

Kelicia91 replied:
기존 `getNumOfLottosCanBuy()` 메소드에서 계산하도록 수정햇고 테스트 코드도 추가햇어요!

minsungryu commented on src/main/java/LottoStore.java:
```diff
@@ -1,33 +1,45 @@
 public class LottoStore {
 
-	public static final String KRW_UNIT = "원";
+	public List<Lotto> sell(LottoPayment payment, List<String> manualLottos) {
```
앞서 말씀드린 것의 연장선으로 `LottoBuyer`나 `LottoStore` 모두 도메인 객체라고 생각이 드는데요
현재는 `LottoBuyer`가 내부에서 View를 통해 `LottoPayment`를 생성하고 있지만
외부에서 생성된 `LottoPayment`를 인자로 받아서 `LottoBuyer`를 생성한 뒤 `LottoStore` 메서드의 인자로 넘겨주면 더 자연스럽지 않을까 생각이 들었습니다.

```java
public List<Lotto> sell(LottoBuyer buyer, List<String> manualLottos) {
    ...
}
```

추가로 `List<Lotto>`도 일급 컬렉션으로 구현해보는 연습을 해봐도 좋을 것 같습니다

Kelicia91 replied:
엄, 일단 먼저 설명을 드리자면 `LottoBuyer`는 도메인이 아니라 `LottoApp`에서 로또 구매를 다루기엔 호흡이 너무 긴것 같아서 따로 분리한 controller 에요(이름이 다소 도메인스럽지만..)
모호하게 보일바엔 차라리 `LottoBuyer` 로직을 `LottoApp.run`에 전부 풀어쓰는게 더 명확할까? 싶은 생각이 드네요. (view도 굳이 나눌 필요없고 store도 주입없이 app의 멤버변수로 직접 접근하면 되니..) 어떻게 생각하시나요 ?_?

minsungryu replied:
아이고 제가 부분만 보다보니 전체를 놓쳤나보네요 ㅎㅎ
다만 이름이 혼란을 줄 수 있다는 점에서 클래스명을 변경해보시는 건 어떨까요?
`Controller` 관점이라면 크게 구조를 변경할 필요는 없어보입니다.

그렇다면 오히려 `LottoApp`이 생성자에서 `View`를 초기화하는 부분이 걸리네요.
`run()` 시점에 `View`를 생성해서 `Controller`에 넘겨주는 방법도 좋을 것 같습니다.
`LottoApp`이 객체 생성의 책임을 갖는게 부담스럽다면 팩토리를 선언해도 괜찮을 것 같네요.

Kelicia91 replied:
> buyer 를 buyLottosController 로 재명명

> LottoApp이 main controller 고, BuyLottosController가 sub controller 격이라 app 에서 view 하나만 생성하고 sub controller 생성시 view를 주입해주는 방식으로 변경했어요!

(앞서 코멘트 남겼듯이) 요렇게 변경해보았어요~ 
팩토리까진 쵸큼(?) 과한 것 같아서^^; app 에서 생성처리 햇어요 🙏

minsungryu commented on src/main/java/ManualLottoAmount.java:
```diff
@@ -0,0 +1,71 @@
+
+	private final int amount;
+
+	private ManualLottoAmount(LottoPayment payment, int amount) {
+		validate(payment, amount);
+		this.amount = amount;
+	}
```
`LottoPayment`와 `amount`를 받는데 `payment`는 검증에 사용되고 사실상 `amount`만 초기화하고 있어서 살짝 어색하게 느껴지는데요,
`LottoPayment`와 `manualLottoCount(?? - 가칭)`를 받아서 초기화하는 방식으로 구현해보는 것은 어떨까요?

Kelicia91 replied:
> LottoPayment와 manualLottoCount(?? - 가칭)를 받아서 초기화하는 방식

요 부분을 제가 이해를 잘 못해서;; 좀더 구체적으로 설명해주실 수 있을까요ㅜ?

minsungryu replied:
이 부분은 `amount`를 제가 잘못이해한 것 같아요.
금액이라고 생각했는데 다시 보니 수동 로또 구매 갯수를 의미하는 것 같네요.
(테스트도 어쩐지 숫자가 너무 작다고 생각했던..)

Kelicia91 replied:
`amount`가 오해의 여지가 잇는 것 같아 `quantity`로 수정햇습니다요

minsungryu reviewed (changes requested):
지우님 5단계까지 미션 진행하시느라 고생 많으셨습니다.
코드의 가독성도 좋고 단위 테스트도 잘 구현해주신 것 같습니다 👍 
그 외 몇 가지 소소한 코멘트들을 남겼는데요
확인해주시면 좋을 것 같습니다 :)

minsungryu commented on src/main/java/ManualLottoAmount.java:
```diff
@@ -0,0 +1,71 @@
+		if (amount < 0) {
+			throw new ManualLottoAmountException(payment);
+		}
+		if (payment.getNumOfLottosCanBuy() < amount) {
```
다만 이 부분에서는 `get()` 후 비교보다는 `payment` 객체에 물어보는 방식의 비교가 더 좋을 것 같습니다.

Kelicia91 replied:
아앗 이런 실수를 수정햇어요!

Kelicia91 committed:
refactor: rename LottoBuyer to BuyLottosController

- refactor: BuyLottosController 생성시점에 view 주입
- refactor: BuyLottosController.buy() to buyLottosAt(store)

Kelicia91 committed:
refactor: LottoPayment 생성자 내 계산로직을 메소드로 분리

- refactor: rename the param KRW to krw

Kelicia91 committed:
refactor: rename ManualLottoAmount to ManualLottoQuantity

Kelicia91 committed:
refactor: ManualLottoQuantity.validate() - 갯수 비교를 직접하지 않고 payment에 위임

Kelicia91 committed:
refactor: List<Lotto>를 Lottos 클래스로 분리

- refactor: Lottos 수동/자동 생성을 LottosFactory 클래스로 분리

Kelicia91 commented:
`List<Lotto>` 컬렉션을 `Lottos`로 분리했고, `Lottos` 수동/자동 생성은 `LottosFactory`로 분리해보았어요. 리뷰 부탁드려요~ 🙏

minsungryu commented on src/main/java/LottoStore.java:
```diff
@@ -1,19 +1,12 @@
-			.collect(Collectors.toList());
+		final Lottos lottosManually = LottosFactory.manuallyFrom(manualLottos);
+		final Lottos lottosAuto = LottosFactory.autoFrom(payment, lottosManually);
+		return Lottos.merge(lottosManually, lottosAuto);
```
객체들이 잘 협력하는 모습이 보이네요!
가독성이 훨씬 좋아진 것 같습니다 👍

minsungryu commented on src/main/java/Lottos.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class Lottos {
```
👏 👏 👏  일급 컬렉션도 멋지게 구현해주셨네요

minsungryu commented on src/main/java/LottosFactory.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class LottosFactory {
```
팩토리까지.. 멋집니다

minsungryu reviewed (approved):
지우님 로또 미션 성공적으로 완료하셨군요! 💯 
바로 merge 하겠습니다.
다음 미션도 화이팅입니다~
